### PR TITLE
sensors/PAG7920: Support PixArt image sensor PAG7920.

### PR DESCRIPTION
**Pull request by Lak4CYUT**
1. Add a sensor driver for PAG7920.
2. Add PAG7920-related definitions.

### CODE REVIEW

**Comment by kwagyeman** [comment]
@Lak4CYUT - I have one of these sensors. I can test this PR after all review comments are addressed from me and @iabdalkader.

File: src/omv/sensors/pag7920_reg.h [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+#define V_R009_QQ_H_ON              (0x02)
+#define V_R009_QQ_H_OFF             (0x00)
+
+#define V_R1C2_Q_V_ON               (0xF7)
+#define V_R1C2_Q_V_OFF              (0x00)
+#define V_R1C2_QQ_V_ON              (0xF5)
+#define V_R1C2_QQ_V_OFF             (0x02)
+
+const uint16_t GRAY_QVGA [] = {
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920_reg.h:100]:
```suggestion
const uint16_t GRAY_QVGA [] = {
```
We typically put register arrays in the `*.c` files, just for consistency.  Also they're typically 2D arrays, with `0x00`, or `0xFF`, marking the end of the array. For example:
```c
static const uint8_t qvga_regs[][2] = {
    { COM3,         0x00 },
    ....
    { 0xFF,         0xFF },
};
....
    for (int i = 0; qvga_regs[i][0] != 0xFF; i++) {
        ret |= omv_i2c_writeb(&sensor->i2c_bus, sensor->slv_addr, regs[i][0], regs[i][1]);
    }
```

File: src/omv/sensors/pag7920.c [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+
+extern uint8_t _line_buf;
+static int8_t g_bank_cache = -1;
+
+static bool g_f_hflip = false;
+static bool g_f_vflip = false;
+
+static int switch_bank(sensor_t *sensor, uint8_t bank) {
+    if (g_bank_cache != bank) {
+#if (SENSOR_I2C_DEBUG == 1)
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:54]:
```suggestion
    #if (SENSOR_I2C_DEBUG == 1)
```
Those should match the indent of the following line, our formatter doesn't automatically fix it.

File: src/omv/sensors/pag7920.c [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+        int res = omv_i2c_writeb(&sensor->i2c_bus, sensor->slv_addr, ta_seq[i],
+                                 ta_seq[i + 1]);
+        if (res) {
+            return res;
+        }
+    }
+#undef ta_seq
+    return 0;
+}
+//////////////////////////////////////////////////////////////////////////////
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:114]:
```suggestion

```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

**Comment by iabdalkader** [comment]
Thanks for sending this driver! Note I have a PAG7920 and I tested it with OPENMV4P, but it's not detected, not sure why.

**Comment by Lak4CYUT** [comment]
> Thanks for sending this driver! Note I have a PAG7920 and I tested it with OPENMV4P, but it's not detected, not sure why.

Not detected? Wow, I have tested with OPENMV4 before submit the PR. I will find a OPENMV4P and verify it again.

File: src/omv/sensors/pag7920.c [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+#if (SENSOR_I2C_DEBUG == 1)
+    printf("R Reg: 0x%02X\r\n", addr);
+#endif
+    return omv_i2c_readb(&sensor->i2c_bus, sensor->slv_addr, addr, p_val);
+}
+
+static int init_sensor(sensor_t *sensor) {
+    write_reg_w_bank(sensor, BANK_0, R_GLOBAL_RESET, V_GLOBAL_RESET);
+    mp_hal_delay_ms(50);
+#define ta_seq GRAY_QVGA
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:99]:
```suggestion
    #define ta_seq GRAY_QVGA
```

File: src/omv/sensors/pag7920.c [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+    int arr_size = sizeof(ta_seq) / sizeof(ta_seq[0]);
+    for (int i = 0; i < arr_size; i += 2) {
+#if (SENSOR_I2C_DEBUG == 1)
+        printf("W Reg: 0x%02X 0x%02X\r\n", ta_seq[i], ta_seq[i + 1]);
+#endif
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:104]:
```suggestion
        #if (SENSOR_I2C_DEBUG == 1)
        printf("W Reg: 0x%02X 0x%02X\r\n", ta_seq[i], ta_seq[i + 1]);
        #endif
```

And so on...

File: src/omv/sensors/pag7920.c [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+        int res = omv_i2c_writeb(&sensor->i2c_bus, sensor->slv_addr, ta_seq[i],
+                                 ta_seq[i + 1]);
+        if (res) {
+            return res;
+        }
+    }
+#undef ta_seq
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:111]:
```suggestion
    #undef ta_seq
```

File: src/omv/sensors/pag7920.c [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+    return 0;
+}
+static int reset(sensor_t *sensor) {
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:114]:
```suggestion

static int reset(sensor_t *sensor) {
```

File: src/omv/sensors/pag7920.c [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+        case FRAMESIZE_QVGA:
+            res |= write_reg_w_bank(sensor, BANK_1, R1_C2,
+                                    enable?V_R1C2_Q_V_ON:V_R1C2_Q_V_OFF);
+            break;
+        case FRAMESIZE_QQVGA:
+            res |= write_reg_w_bank(sensor, BANK_1, R1_C2,
+                                    enable?V_R1C2_QQ_V_ON:V_R1C2_QQ_V_OFF);
+            break;
+        default:
+            printf("Unsupported.\r\n");
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:470]:
```suggestion
```

File: src/omv/sensors/pag7920.c [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+
+static int switch_bank(sensor_t *sensor, uint8_t bank) {
+    if (g_bank_cache != bank) {
+    #if (SENSOR_I2C_DEBUG == 1)
+        printf("W Reg: 0x%02X 0x%02X\r\n", REG_BANK, bank);
+#endif
+        int res = omv_i2c_writeb(&sensor->i2c_bus, sensor->slv_addr, REG_BANK, bank);
+        if (res) {
+            g_bank_cache = -1;  // Encountered an unknown error, clear cache.
+            printf("switch bank failed, res = %d\n", res);
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:60]:
```suggestion
            #if (SENSOR_I2C_DEBUG == 1)
            printf("switch bank failed, res = %d\n", res);
            #endif
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

**Comment by Lak4CYUT** [comment]
> I did, but it's still not detected. I can see the registers dump if I enable the debug printf, so I think the sensor I have is working. Maybe @kwagyeman can test with the RT1060.

Oops, I will check it again with a clean build, but I think the source code should be fine.
I am concerned that the sensor board you own may need a check.

By the way, how can I fetch the code that is 100% in sync with you tested? Is there a commit hash I can use? 
(Sorry, I am not very familiar with Github)

> BTW re the debug print statements, we already have something like that built-in you can just use `debug_printf` if you like. It's disabled by default, and can be enabled with:
> 
> ```c
> make -j12 TARGET=OPENMV4 DEBUG_PRINTF=1
> ```
Sure, I would very much like to follow your design. I will update my commit to fix it.

**Comment by iabdalkader** [comment]
> Oops, I will check it again with a clean build, but I think the source code should be fine.
> I am concerned that the sensor board you own may need a check.

No worries, maybe Kwabena can test both the H7 and RT later today.

**Comment by Lak4CYUT** [comment]
I have made a clean build using the following git command, it should align with your test environment.
```
git -c protocol.version=2 fetch --no-tags --prune --no-recurse-submodules --depth=1 upstream +597701312eac7bba9e4881a06645b7dd92abaa13:refs/remotes/pull/2199/merge
git checkout --progress --force refs/remotes/pull/2199/merge
make clean TARGET=OPENMV4 -C src
make TARGET=OPENMV4 -C src
```
It is works fine for me with OPENMV4. 
We will discuss internally how to determine whether this issue is related to the sensor board.
FYI.

**Comment by kwagyeman** [comment]
Will test soon.

**Comment by kwagyeman** [comment]
Hi, I tested the device on the RT1062 and OpenMV Cam H7 Plus and it works on both.

**Comment by Lak4CYUT** [comment]
> Hi, I tested the device on the RT1062 and OpenMV Cam H7 Plus and it works on both.

Sounds good, so does this mean our driver is ok now?
It seems is this PR contains too many modifications, I prefer to abandon it. 
I will correct all coding style issues and resend a PR, is that ok?

**Comment by kwagyeman** [comment]
Hi, you just force push update your branch. The PR will then be automatically be updated.

**Comment by Lak4CYUT** [comment]
> Hi, you just force push update your branch. The PR will then be automatically be updated.

Ok, I would try it.

**Comment by Lak4CYUT** [comment]
@kwagyeman 
It is seems your coding style checking system doesn't agree your suggestion XD.

That
`#define ta_seq GRAY_QVGA`
is useful for my internal development, can I keep it?

File: src/omv/sensors/pag7920.c [hunk 10]
```diff
@@ -0,0 +1,654 @@
...
+#include "pag7920_reg.h"
+
+#define SENSOR_I2C_DEBUG (0)
+
+#define US (1000000)  // microsecond
+#ifndef M_LN29
+#define M_LN29  (3.36729582998647402718)
+#endif
+
+const uint16_t GRAY_QVGA [][2] = {
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:32]:
```suggestion
// Grayscale/QVGA
static const uint16_t default_regs[][2] = {
```

File: src/omv/sensors/pag7920.c [hunk 11]
```diff
@@ -0,0 +1,654 @@
...
+            val_R_WOI_VStart = 9;
+            val_R_ABC_grad1 = 120;
+            val_R2_D9_B0 = 1;
+            val_R009_H = g_f_hflip ? V_R009_QQ_H_ON : V_R009_QQ_H_OFF;
+            val_R2_D9_B1 = 0;
+            val_R_AE_Size__div4_X = 40;
+            val_R_AE_Size__div4_Y = 30;
+            break;
+        default:
+            printf("Unsupported resoultion.\n");
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:345]:
```suggestion
```

File: src/omv/sensors/pag7920.c [hunk 12]
```diff
@@ -0,0 +1,654 @@
...
+    static const uint32_t digital = (OMV_PAG7920_XCLK_FREQ / US) / 2;
+    uint32_t clk_num;
+    read_reg_w_bank(sensor, BANK_4, Reg_ExpPxclkNum_0, (uint8_t *) &clk_num);
+    read_reg_w_bank(sensor, BANK_4, Reg_ExpPxclkNum_1, (uint8_t *) (&clk_num) + 1);
+    read_reg_w_bank(sensor, BANK_4, Reg_ExpPxclkNum_2, (uint8_t *) (&clk_num) + 2);
+    read_reg_w_bank(sensor, BANK_4, Reg_ExpPxclkNum_3, (uint8_t *) (&clk_num) + 3);
+    clk_num &= 0x0fffffff;
+#ifdef DEBUG_AE
+    printf("Reg_ExpPxclkNum = %ld\n", clk_num);
+#endif
```
Comment (iabdalkader) [inline src/omv/sensors/pag7920.c:550]:
```suggestion
    debug_printf("Reg_ExpPxclkNum = %ld\n", clk_num);
```

**Review by iabdalkader — CHANGES_REQUESTED** [review]

**Comment by iabdalkader** [comment]
> It seems your coding style checking bot doesn't agree with your suggestion, XD.

Yes you're right! You can revert that line to make the test pass.

> `#define ta_seq GRAY_QVGA`
> is useful for my internal development, can I keep it?

Sure.

**Comment by iabdalkader** [comment]
> It seems is this PR contains too many modifications, I prefer to abandon it.

Don't worry about the commits, I will squash the PR when merging.

**Pull request merged**